### PR TITLE
ci: refactor node controller unit tests

### PR DESCRIPTION
**Pull request by m-ildefons**
Refactor node controller unit tests:
- Break up single test function into multiple test functions
- Create a clear separation between test case input data and expected result data
- Implement existing tests with new structure as individual test functions
- Avoid checking things that aren't subject matter of the test case at hand

related-to: longhorn/longhorn#7332

### CODE REVIEW

File: controller/node_controller_test.go [hunk 1]
```diff
@@ -2,132 +2,169 @@ package controller
...
+	nodeIndexer cache.Indexer
+
+	controller *NodeController
 }
 
-func fakeTopologyLabelsChecker(kubeClient clientset.Interface, vers string) (bool, error) {
-	return false, nil
+// This data type contains resource that exist in the cluster environment, like
+// nodes and pods
+type TestCase struct {
```
Comment (PhanLe1010) [inline controller/node_controller_test.go:78]: Should we name this struct to `NodeTestCase`? I think if we adopt this new test patent for other controllers in this package, the generic `TestCase` naming will be conflicted later on

File: controller/node_controller_test.go [hunk 2]
```diff
@@ -621,233 +1017,349 @@ func (s *TestSuite) TestSyncNode(c *C) {
...
+		condition.LastTransitionTime = ""
+		condition.Message = ""
+		node.Status.Conditions[idx] = condition
+	}
+	c.Assert(node.Status.Conditions, DeepEquals, exp.nodeStatus[node.Name].Conditions)
+}
 
-		nIndexer := informerFactories.LhInformerFactory.Longhorn().V1beta2().Nodes().Informer().GetIndexer()
-		pIndexer := informerFactories.KubeInformerFactory.Core().V1().Pods().Informer().GetIndexer()
+func (s *NodeControllerSuite) checkDiskConditions(c *C, tc *TestCase, exp *Expectation, node *longhorn.Node) {
```
Comment (PhanLe1010) [inline controller/node_controller_test.go:1225]: Looks like we don't call this function in any the test cases. Is it expected?

File: controller/node_controller_test.go [hunk 3]
```diff
@@ -2,132 +2,169 @@ package controller
...
-	return map[string]*corev1.Node{
-		TestNode1: kubeNode1,
-		TestNode2: kubeNode2,
-	}
+// This data type contains expected results in the form of resources. Each test
+// will set up fixtures, input resources and expected results. Then it will
+// initialize the mock API with initTest, execute a function of the controller
+// and finally execute a set of assertions, which compare the actual contents of
+// the mock API to the expected results
+type Expectation struct {
```
Comment (PhanLe1010) [inline controller/node_controller_test.go:93]: Same as above, should we chane the name to `NodeControllerTestExpectation`?

**Review by PhanLe1010 — APPROVED** [review]
I very like the new test design pattern! 
In general LGTM expect for some small nits

File: controller/node_controller_test.go (hunk 1 above)
Comment (m-ildefons) [reply controller/node_controller_test.go]: I'll rename it to `NodeControllerFixture` and `NodeControllerExpectation` respectively. If the other test suites also match this structure, we don't need to maintain separate fixture and expectation data structures for each controller, but until I've looked at other tests this makes a lot of sense.

File: controller/node_controller_test.go (hunk 2 above)
Comment (m-ildefons) [reply controller/node_controller_test.go]: Good spot, this needs to be called for all tests that have non-empty disk conditions in their expectations.

**Review by PhanLe1010 — APPROVED** [review]
LGTM
